### PR TITLE
168 active learning pip package endpoints me

### PR DESCRIPTION
**Pull request by MalteEbner**
closes: https://github.com/lightly-ai/lightly-core/issues/168, https://github.com/lightly-ai/lightly/pull/136
- added generate files for the endpoints
- added a scorer for classification with tests
- added support for sampling with active learning scores

HOW TO TEST:
- run the script `test_al_step_3.py`

### CODE REVIEW

File: lightly/active_learning/scorers/classification.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+
+def entropy(probs: np.ndarray, axis: int = 1):
+    zeros = np.zeros_like(probs)
+    log_probs = np.log2(probs, out=zeros, where=probs > 0)
+    entropies = -1 * np.sum(probs * log_probs, axis=axis)
+    return entropies
+
+
+class ScorerClassification(Scorer):
```
Comment (IgorSusmelj) [inline lightly/active_learning/scorers/classification.py:15]: Some docstrings here would be helpful. Is this a scorer for "image classification"? Will there be another one for object detection?

File: lightly/active_learning/scorers/classification.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    def __init__(self, model_output: np.ndarray):
```
Comment (IgorSusmelj) [inline lightly/active_learning/scorers/classification.py:16]: Please describe the expected shape from `model_output` in the docstring. Are we talking about something like this?
`NxC` meaning we have N samples and C classes. 
Do we care about the range of the values? Should they be in the range of [0,1] or can they be logits?

**Review by IgorSusmelj — CHANGES_REQUESTED** [review]
I would like documentation on some of the functions where it's not clear whether there are any expectations regarding input/output.

File: lightly/active_learning/scorers/classification.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
+from typing import *
+
+import numpy as np
+
+from lightly.active_learning.scorers.scorer import Scorer
+
+
+def entropy(probs: np.ndarray, axis: int = 1):
```
Comment (philippmwirth) [inline lightly/active_learning/scorers/classification.py:8]: Docstrings, always ;)

Comment (philippmwirth) [reply lightly/active_learning/scorers/classification.py:8]: Also, make it private by prefixing with _

File: lightly/active_learning/scorers/classification.py (hunk 2 above)
Comment (philippmwirth) [reply lightly/active_learning/scorers/classification.py:16]: In the long run we should probably support [0, 1] and logits but let's choose one to start with

File: lightly/api/api_workflow_sampling.py [hunk 4]
```diff
@@ -37,7 +40,12 @@ def sampling(self, sampler_config: SamplerConfig, al_scores: Dict[str, List[int]
 
         # upload the active learning scores to the api
         if al_scores is not None:
-            raise NotImplementedError  # TODO: fill out in later branches
+            if preselected_tag_id is None:
+                raise ValueError
+            for score_type, score_values in al_scores.items():
```
Comment (philippmwirth) [inline lightly/api/api_workflow_sampling.py:45]: I'd suggest to make a check in the pip package whether the number of scores coincides with the size of the tag they're being uploaded to. @MalteEbner @IgorSusmelj thoughts on that?

File: tests/UNMOCKED_end2end_tests/test_al_step3.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+from lightly.openapi_generated.swagger_client import SamplingMethod
+
+from lightly.utils import save_embeddings
+
+from lightly.active_learning.config.sampler_config import SamplerConfig
+
+
+def t_est_unmocked_complete_workflow(path_to_dataset: str, token: str, dataset_id: str):
+    # define the api_client and api_workflow
+    api_workflow_client = ApiWorkflowClient(host="https://api-dev.lightly.ai", token=token, dataset_id=dataset_id)
```
Comment (philippmwirth) [inline tests/UNMOCKED_end2end_tests/test_al_step3.py:22]: Can you switch to getting the host from the env variable? Also in the other unmocked script because I don't feel so comfortable exposing our developer api to everyone.

File: tests/active_learning/test_ScorerClassification.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+        predictions_normalized = predictions / np.sum(predictions, axis=1)[:, np.newaxis]
+        model_output = predictions_normalized
+        scorer = ScorerClassification(model_output)
+        scores = scorer._calculate_scores()
+        scores_prediction_entropy = scores["prediction-entropy"]
+        scores_prediction_margin = scores["prediction-margin"]
+
+        assert scores_prediction_entropy.shape == (n_samples, )
+        assert scores_prediction_margin.shape == (n_samples, )
+        assert all(scores_prediction_entropy > 0)
```
Comment (philippmwirth) [inline tests/active_learning/test_ScorerClassification.py:22]: This assertion will fail for the following example input:
```
[
   [1.0, 0.0, 0.0, 0.0, 0.0]
]
```
because the logarithm retains the original value for all values <= 0.

**Review by philippmwirth — CHANGES_REQUESTED** [review]
Please add docstrings to all the functions. If you think they are self-explanatory you can simply add a single sentence like so:
```Python
def my_fun():
    """This is my function.
    
    """
```

Otherwise it's looking good! Great work @MalteEbner!

File: tests/UNMOCKED_end2end_tests/test_al_step3.py (hunk 5 above)
Comment (MalteEbner) [reply tests/UNMOCKED_end2end_tests/test_al_step3.py:22]: I would not call it "exposing" the api, as it will be accessible from everywhere even if we do not give the easily-to-guess URL in the pip package. I someone really wants to get to it, trying out "api-dev" is not really harder than searching our pip-package for some URLs. The URL is protected by a token anyways. In my eyes removing the url is more like preventing burglary on your house by removing your address from the telephone book.
If you really want me to, I could change that, but I do not see it as a security measure.

Comment (philippmwirth) [reply tests/UNMOCKED_end2end_tests/test_al_step3.py:22]: I don't see it as a security threat, more as a bad practice but focus on the important stuff first

File: lightly/api/api_workflow_sampling.py (hunk 4 above)
Comment (MalteEbner) [reply lightly/api/api_workflow_sampling.py:45]: I did that in the ActiveLearningAgent.query()

File: lightly/active_learning/scorers/classification.py (hunk 1 above)
Comment (MalteEbner) [reply lightly/active_learning/scorers/classification.py:29]: Yes, there will be another one for object detection in the future, but first we need to define the formats of the prediction of an  object detection model  which will be supported.

File: lightly/active_learning/scorers/classification.py (hunk 2 above)
Comment (MalteEbner) [reply lightly/active_learning/scorers/classification.py:42]: @philippmwirth I do not know how to calculate the prediction entropy and prediction margin with logits, do you know any literature on that?

File: tests/UNMOCKED_end2end_tests/test_al_step3.py (hunk 5 above)
Comment (MalteEbner) [reply tests/UNMOCKED_end2end_tests/test_al_step3.py:22]: I would keep it there as it means one less error which could be made while testing.

File: lightly/active_learning/scorers/classification.py (hunk 2 above)
Comment (philippmwirth) [reply lightly/active_learning/scorers/classification.py:43]: Check out the documentation on the cross entropy loss [here](https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html). But again, it's ok to add this later as long as we specify the input format clearly.

File: lightly/api/api_workflow_sampling.py (hunk 4 above)
Comment (philippmwirth) [reply lightly/api/api_workflow_sampling.py:45]: Didn't see that, thanks for the hint

File: lightly/active_learning/scorers/classification.py (hunk 1 above)
Comment (MalteEbner) [reply lightly/active_learning/scorers/classification.py:29]: I created an issue for it here: https://github.com/lightly-ai/lightly/issues/137

File: lightly/active_learning/scorers/classification.py (hunk 2 above)
Comment (MalteEbner) [reply lightly/active_learning/scorers/classification.py:43]: So the cross entropy loss out of logits is basically computed by first applying a softmax, as it computes the `exp(x_i)/sum(exp(x))`. The softmax gives the probability distribution as we want it. Thus our users would need to run the logits through a softmax first before setting the model_output in the scorer.

Comment (philippmwirth) [reply lightly/active_learning/scorers/classification.py:43]: Yes, let's do it this way for the moment. The advantage from calculating the entropy from logits is improved numerical stability

**Review by philippmwirth — APPROVED** [review]

**Review by IgorSusmelj — APPROVED** [review]
